### PR TITLE
feat: add support for configuring network protocol_id's

### PR DESCRIPTION
**Pull request by KolbyML**
### What was wrong?
We aren't able to configure protocol_id in the case we want to support different networks like a testnet
### How was it fixed?
This is basically the same thing we already did except we use a bidirectional map instead of trait conversions acting as a bidirectional map.

For storing mainnet/future network data we will use a folder structure I seen L1 clients use ``trin/mainnet/...`` ``trin/futurenetworkname/...``

Now we can easily configure future networks

### CODE REVIEW

**Review by njgheorghita — COMMENTED** [review]
I appreciate the pr & work towards this direction. But imo we shouldn't merge any testnet implementation until https://github.com/ethereum/portal-network-specs/pull/258 is approved by all client teams & merged

**Comment by KolbyML** [comment]
@njgheorghita I removed the 10 lines relating to a testnet you can review my code now

**Comment by KolbyML** [comment]
By what is said between you and piper this is in a deadlock now. This PR more has to do with configuring protocol ID's and I don't see how that a PR formalizing protocol_ids for a testnet blocks a pr allowing the option to configure protocol ID's.

Being able to configure network parameters is a standard feature in all L1 clients. I can comprehend how we wouldn't want to formalize the testnet protocol_ids in our code, but that is 15 lines of 200. This code could have been reviewed and we could have just removed the 15 line testnet part instead of a blanket ignore because the word testnet was mentioned.

Shouldn't by the same logic we should have blocked TraceGossip till it was reviewed into the spec and reviewed by all client teams? I don't see why 95% (configure protocol id's) of this PR is being blocked for 5% (15 lines of testnet parameters)

**Comment by pipermerriam** [comment]
I wouldn't categorize my opinion as "strong".  Only that since we haven't done "testnet" yet my intuition is that actually doing the implementation will likely uncovers somes "unknown unknown" types of learning... but feel free to proceed however you guys feel is best.

I think the only thing I have a "strong" opinion on is that we **should not** write the spec and then blindly implement the clients to follow the spec.  Lets make sure that client implementation learnings are actively used to influence the specification because things don't always unfold the way we expect them to.

**Review by njgheorghita — COMMENTED** [review]
I think kurtosis could be useful in testing this.
Something like...
- Spin up 3 trin nodes running testnet
- Spin up 3 trin nodes running mainnet
- Send a ping between the testnet & mainnet clients
- Spin up a bridge gossiping testnet data (eg epoch # 1)
- Spin up a bridge gossiping mainnet data (eg epoch # 2)

Test that the mainnet nodes only store epoch # 2 data, and testnet nodes only have epoch # 1 data.

**Comment by KolbyML** [comment]
@njgheorghita @ogenev ready for a review

I added the small peertest I talked about in the call and comment above

File: ethportal-api/src/types/cli.rs [hunk 1]
```diff
@@ -299,6 +311,14 @@ pub fn check_private_key_length(private_key: &str) -> Result<B256, String> {
     ))
 }
 
+fn network_parser(network_string: &str) -> Result<Arc<NetworkSpec>, String> {
+    match network_string {
+        "mainnet" => Ok(MAINNET.clone()),
+        "testnet" => Ok(TESTNET.clone()),
+        _ => Err(format!("Not a valid network: {network_string}")),
```
Comment (njgheorghita) [inline ethportal-api/src/types/cli.rs:318]: `"Not a valid network: {network_string}, must be 'testnet' or 'mainnet'"`

File: ethportal-api/src/types/portal_wire.rs [hunk 2]
```diff
@@ -174,22 +176,41 @@ pub enum ProtocolId {
...
+        network_name: "mainnet".to_string(),
+    }
+    .into()
+});
+
+pub static TESTNET: Lazy<Arc<NetworkSpec>> = Lazy::new(|| {
+    let mut portal_networks = BiHashMap::new();
+    portal_networks.insert(ProtocolId::State, "0x501A".to_string());
+    portal_networks.insert(ProtocolId::History, "0x501B".to_string());
+    portal_networks.insert(ProtocolId::TransactionGossip, "0x501C".to_string());
```
Comment (njgheorghita) [inline ethportal-api/src/types/portal_wire.rs:204]: I don't see any reason to include `TransactionGossip` & `CanonicalIndices`?

File: ethportal-api/src/types/portal_wire.rs [hunk 3]
```diff
@@ -590,21 +611,20 @@ impl From<Accept> for Value {
...
-        ProtocolId::from_str(hex).unwrap();
+        assert!(!MAINNET.portal_networks.contains_right(hex));
     }
 
     #[test]
     fn protocol_id_encoding() {
         let hex = "0x500A";
-        let protocol_id = ProtocolId::from_str(hex).unwrap();
-        let expected_hex = hex_encode_upper(Vec::try_from(protocol_id).unwrap());
+        let protocol_id = MAINNET.portal_networks.get_by_right(hex).unwrap();
```
Comment (njgheorghita) [inline ethportal-api/src/types/portal_wire.rs:626]: I'm not a huge fan of using the native `BiHashMap` api. It's not obvious (without looking up the map &/or `BiHashMap` source code), what these functions do (aka left vs. right)... Could you `impl` something more intuitive, like `get_protocol_id` / `get_protocol_id_hex` (or whatever fn name makes the most sense here)?

File: portalnet/src/events.rs [hunk 4]
```diff
@@ -118,10 +122,13 @@ impl PortalnetEvents {
 
     /// Dispatch Discv5 TalkRequest event to overlay networks or uTP socket
     fn dispatch_discv5_talk_req(&self, request: TalkRequest) {
-        let protocol_id = ProtocolId::from_str(&hex_encode_upper(request.protocol()));
+        let protocol_id = self
+            .network_spec
+            .portal_networks
+            .get_by_right(&hex_encode_upper(request.protocol()));
```
Comment (njgheorghita) [inline portalnet/src/events.rs:128]: Same here. Some custom `impl` that has a more intuitive name would be better

File: src/lib.rs [hunk 5]
```diff
@@ -100,7 +111,11 @@ pub async fn run_trin(
         beacon_event_tx,
         beacon_jsonrpc_tx,
         beacon_event_stream,
-    ) = if trin_config.networks.iter().any(|val| val == BEACON_NETWORK) {
+    ) = if trin_config
+        .portal_subnetworks
```
Comment (njgheorghita) [inline src/lib.rs:115]: Think we could also do a custom `trin_config.portal_subnetwork.contains()` or something similar

File: tests/self_peertest.rs [hunk 6]
```diff
@@ -206,13 +206,22 @@ async fn peertest_state_offer_contract_bytecode() {
 #[tokio::test(flavor = "multi_thread")]
 #[serial]
 async fn peertest_state_recursive_gossip() {
-    let (peertest, target, handle) = setup_peertest().await;
+    let (peertest, target, handle) = setup_peertest(None).await;
```
Comment (njgheorghita) [inline tests/self_peertest.rs:209]: I think it would be a bit better to get rid of the `Option` and just use `"mainnet"` as the arg here

**Review by njgheorghita — APPROVED** [review]
Looks pretty good, just a few comments on fn naming. But, I'd also like to hold off on merging this until after the db update is deployed, since this will require some changes to the cluster repo, and it'd be better to do the 2 somewhat significant changes in separate goes to isolate their effects/bugs from each other

**Comment by KolbyML** [comment]
> since this will require some changes to the cluster repo,

This PR requires 0 changes to the cluster repo, to my knowledge? Would it be possible to provide an explanation on how that conclusion was achieved? Maybe I am misunderstanding something.

File: ethportal-api/src/types/portal_wire.rs (hunk 2 above)
Comment (KolbyML) [reply ethportal-api/src/types/portal_wire.rs:204]: I mean it is included in our mainnet code

**Comment by KolbyML** [comment]
ready for another look

**Comment by njgheorghita** [comment]
Well, it's not a major change, but we'll need to update the `--networks` flag to `--portal-subnetworks`. Anyways, I guess it's not a huge blocker for this pr, just something that we'll make sure we want to do in parallel. And if we forget to do it, the nodes will crash, so the problem will be pretty obvious. So, I guess nvm let's go ahead with the merge here.

**Pull request merged**